fs-cache/src/memory_limited_storage.rs

```diff
@@ -0,0 +1,273 @@
+                    Ok(value) => {
+                        let actual_size = Self::estimate_size(&value);
+                        if loaded_bytes + actual_size <= self.max_memory_bytes {
+                            self.memory_cache.insert(key, value);
+                            loaded_bytes += actual_size;
+                        }
+                    }
+                    Err(err) => {
+                        log::warn!(
+                            "{} cache: failed to load key={}: {}",
+                            self.label,
+                            key,
+                            err
+                        );
+                    }
+                }
+            }
+        }
+
+        self.current_memory_bytes = loaded_bytes;
+
+        log::debug!(
+            "{} loaded {}/{} bytes in memory",
+            self.label,
+            self.current_memory_bytes,
+            total_bytes
+        );
+
+        Ok(())
+    }
+
+    fn estimate_size(value: &V) -> usize {
+        serde_json::to_vec(value)
+            .map(|v| v.len())
+            .unwrap_or(0)
+    }
+
+    // Write a single value to disk
+    fn write_value_to_disk(&mut self, key: &K, value: &V) -> Result<()> {
+        let file_path = self.path.join(format!("{}.json", key));
+        let mut file = File::create(&file_path)?;
+        file.write_all(serde_json::to_string_pretty(&value)?.as_bytes())?;
+        file.flush()?;
+
+        let new_timestamp = SystemTime::now();
+        file.set_modified(new_timestamp)?;
+        file.sync_all()?;
+
+        Ok(())
+    }
+
+    // Load a single value from disk
+    fn load_value_from_disk(&self, key: &K) -> Result<V> {
+        let file_path = self.path.join(format!("{}.json", key));
+        let file = File::open(&file_path)?;
+        let value: V = serde_json::from_reader(file).map_err(|err| {
+            ArklibError::Storage(
+                self.label.clone(),
+                format!("Failed to read value for key {}: {}", key, err),
+            )
+        })?;
+        Ok(value)
+    }
+
+    fn add_to_memory_cache(&mut self, key: K, value: V) {
+        let value_size = Self::estimate_size(&value);
+
+        // If single value is larger than total limit, just skip memory caching
+        if value_size > self.max_memory_bytes {
+            log::debug!(
+                "{} cache: value size {} exceeds limit {}",
+                self.label,
+                value_size,
+                self.max_memory_bytes
+            );
+            return;
+        }
+
+        // Remove oldest entries until we have space for new value
+        while self.current_memory_bytes + value_size > self.max_memory_bytes
+            && !self.memory_cache.is_empty()
+        {
+            if let Some((_, old_value)) = self.memory_cache.pop_front() {
+                self.current_memory_bytes = self
+                    .current_memory_bytes
+                    .saturating_sub(Self::estimate_size(&old_value));
```

Comment:
Now I understand, why you've defined this function. But I think we can store sizes in the in-memory mapping next to the values themselves. That would be precise and faster. We can consider the memory overhead as fair trade-off.

Reply:
I would also add `debug_assert` that `current_memory_bytes  - old_value >= 0`.